### PR TITLE
Add custom APIObject serializer and de-serializer for pickle module

### PR DESCRIPTION
**Pull request by JaurbanRH**
The pickle module implements binary protocols for serializing and de-serializing a Python object structure and with the default serializer it wasn't possible to serialize APIObjects, therefore it is necessary to create a custom one.

### CODE REVIEW

File: packages/openshift/apiobject.py [hunk 1]
```diff
@@ -885,3 +885,25 @@ def execute(self, cmd_to_exec=None, stdin=None, container_name=None, auto_raise=
...
+
+    def __setstate__(self, state):
+        """
+        Custom de-serializing for pickle module.
+        The pickle module implements binary protocols for serializing and de-serializing a Python object structure.
+        More info here: https://docs.python.org/3/library/pickle.html#object.__setstate__
+        """
+        with state["context"]:
+            result = selector(f"{state['kind']}/{state['name']}").object_json()
+            super().__init__(string_to_model=result)
```
Comment (bradmwilliams) [inline packages/openshift/apiobject.py:909]: What are you planning on doing with the pickle files that get created from this change?  
I ask, because "Context" is just a programmatic construct that's completely dependent on `oc`'s context  (`oc config get-contexts`) from the underlying environment.  For example, if you have multiple clusters (A and B) and generate pickle files from resources on cluster A.  The information in the pickled files can only be used to locate a resource of `kind/name` on the currently configured `context` of `oc` and not actually anything that's physically stored in the pickled file.  So, in this scenario, that could literally be cluster A or B.

Basically, the context isn't anything that we would, traditionally, want to serialize here.  The "important" bit of information is in the APIObject's Model itself.

Comment (bradmwilliams) [inline packages/openshift/apiobject.py:909]: > super().__init__(string_to_model=result)

This code will throw an Exception.

**Review by bradmwilliams — CHANGES_REQUESTED** [review]

File: packages/openshift/apiobject.py (hunk 1 above)
Comment (JaurbanRH) [reply packages/openshift/apiobject.py:909]: During unpickling of APIObject I am getting this error hence I need a custom serializer.
```
 File "plugin.py", line 111, in pytest_sessionfinish
    self.merge_dicts(pickle.load(f))
                     ^^^^^^^^^^^^^^
  File "/openshift/model.py", line 280, in __setitem__
    super(Model, self).__setitem__(key, to_model_or_val(value, case_insensitive=self.__case_insensitive))
                                                                                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/openshift/model.py", line 252, in __getattr__
    raise AttributeError
AttributeError
```
We are using this library for testing of product deployed on openshift. We are always using only one cluster at the time and the expected usecase is that we run our testsuite that creates some entities at the openshift, we store information about this entities with pickle, upgrade those entities to next version and then load those information and rerun our tests with the same entities.

Comment (JaurbanRH) [reply packages/openshift/apiobject.py:909]: Yes, you are right, there should be `self` instead of `super()`.

**Comment by bradmwilliams** [comment]
/ok-to-test
@JaurbanRH I've applied the label to test, but it's going to fail because you'll need to run the `./ansible/rebuild_module.sh` locally and include the updated files.

**Review by bradmwilliams — APPROVED** [review]

**Review by hoxhaeris — APPROVED** [review]

**Comment by bradmwilliams** [comment]
@JaurbanRH Your changes have been packaged and released in version `1.0.19`
